### PR TITLE
Save TF State as assets

### PR DESCRIPTION
**Pull request by ChrisAmora**
### Summary

Move terraform state management to Twilio Cloud function service.

### Checklist

- [x] Tested changes end to end
- [x] Updated documentation
- [x] Requested one or more reviewers

### CODE REVIEW

File: infra-as-code/state/pull_state.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/bin/bash
+set -e
+echo "Trying to import Terraform State for $ENVIRONMENT"
+echo "### Job Results "
+
+IFS="|" read -ra tf_state_files <<<"$TF_STATE_FILES"
+
+tfstate_bucket=$(twilio api:serverless:v1:services:fetch --sid tfstate -o json | jq -c '.[].domainBase // empty' | sed 's/"//g')
+
+if [ -n "$tfstate_bucket" ]; then
```
Comment (jhunter-twilio) [inline infra-as-code/state/pull_state.sh:10]: i realize it complicates things a little more but do you think we could store and retrieve these as protected assets?

File: .github/workflows/terraform_deploy.yml
Comment (jhunter-twilio) [inline .github/workflows/terraform_deploy.yml:1]: it seems this was renamed from .yaml to .yml - if theres no particular reason do you think you could rename it back?

File: infra-as-code/state/deploy-function/functions/deploy.js
Comment (jhunter-twilio) [inline infra-as-code/state/deploy-function/functions/deploy.js:1]: would it be easier to just delete the service and get rid of this? 

`twilio api:serverless:v1:services:remove --sid ZS.....`

if that is not possible for whatever reason can we add a comment in here explaining this function is really just used as a method of "deleting" the tfstate by blasting over the top of the service with this function?

**Review by jhunter-twilio — CHANGES_REQUESTED** [review]
@ChrisAmora - this is an excellent bit of work - i am excited to move it into the main branch.  I have some thoughts for minor improvements before we do that.  Please let me know what you think.

File: infra-as-code/state/pull_state.sh (hunk 1 above)
Comment (ChrisAmora) [reply infra-as-code/state/pull_state.sh:10]: I tried to do that, but I didn't find a way to download the asset.
Here, it says that only Twilio can download a protected asset.
image
I tried to do curl -u "$TWILIO_ACCOUNT_SID:$TWILIO_AUTH_TOKEN" -o - $url but it didn't work too :( 
Do you know some other way to do that? Thank you.

File: .github/workflows/terraform_deploy.yml
Comment (ChrisAmora) [reply .github/workflows/terraform_deploy.yml:1]: I just renamed it back!

File: infra-as-code/state/deploy-function/functions/deploy.js
Comment (ChrisAmora) [reply infra-as-code/state/deploy-function/functions/deploy.js:1]: This is the function that I use to deploy the assets. I will document it!

File: infra-as-code/state/pull_state.sh (hunk 1 above)
Comment (jhunter-twilio) [reply infra-as-code/state/pull_state.sh:10]: followed up directly.

Comment (ChrisAmora) [reply infra-as-code/state/pull_state.sh:11]: FIxed. Bui we will need the TWILIO_AUTH_TOKEN now!

Comment (dremin) [reply infra-as-code/state/pull_state.sh:11]: Is it possible to do this without TWILIO_AUTH_TOKEN? I'm not sure we want to promote using the auth token in CI.

Comment (dremin) [reply infra-as-code/state/pull_state.sh:11]: What if we use a private asset, but use a public function to access it? The public function could accept the API key/secret and validate it before returning the asset data.

Comment (dremin) [reply infra-as-code/state/pull_state.sh:11]: Here is an example of exactly this! 😄 https://github.com/twilio-professional-services/flex-project-template/blob/main/serverless-schedule-manager/functions/admin/fetch-config.js

**Review by lvaladares — CHANGES_REQUESTED** [review]
can we start moving the pipelines to makefile structure? need it to add serverless deploys

**Comment by ChrisAmora** [comment]
> can we start moving the pipelines to makefile structure? need it to add serverless deploys

Not in this PR. The objective here is to change the deployment pipeline as seamlessly as possible.

File: .github/workflows/flex_deploy.yaml
Comment (dremin) [inline .github/workflows/flex_deploy.yaml:1]: Can you remove the spaces that got added before the `# FEATURE` and `# END FEATURE` lines? These are used by scripting to remove blocks of code and I think this spacing will result in invalid yaml if not removed.

**Review by lvaladares — APPROVED** [review]

File: .github/workflows/flex_deploy.yaml
Comment (ChrisAmora) [reply .github/workflows/flex_deploy.yaml:1]: Sure! it was the auto formatter.

Comment (ChrisAmora) [reply .github/workflows/flex_deploy.yaml:1]: I didn't know it was used for that too. Good to know.

**Review by jhunter-twilio — APPROVED** [review]
this looks great and puts us in a much better spot for improving the "import" stage!! Great work! thank you for the tweaks!

**Pull request merged**